### ISSUE TITLE
empty_enum_arguments false positive with static func

### ISSUE THREAD

**Issue by keith** [realm/SwiftLint#3294]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

With this code:

```swift
struct Foo: Equatable {
    static func empty() -> Foo {
        return Foo()
    }
}

func bar() {
    let foo = Foo()
    if foo == .empty() {
        print("empty")
    }
}
```

And this config:

```yaml
whitelist_rules:
  - empty_enum_arguments
```

`.empty()` is falsely marked as something where `()` can be removed. But those are required to build:

```
foo.swift:9:16: error: member 'empty()' is a function; did you mean to call it?
    if foo == .empty {
               ^
                    ()
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
swiftlint lint --config foo.yml --path foo.swift
```

```
foo.swift:9:21: warning: Empty Enum Arguments Violation: Arguments can be omitted when matching enums with associated types if they are not used. (empty_enum_arguments)
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.40.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? portable_swiftlint.zip
* Paste your configuration file: above

related to https://github.com/realm/SwiftLint/issues/3103 and https://github.com/realm/SwiftLint/issues/3122

**Comment by keith** [comment]
cc @marcelofabri @lordzsolt

**Comment by jpsim** [comment]
This is a regression from 0.39.2, which did not report any violations with the provided repro steps.

**Comment by jpsim** [comment]
Introduced in https://github.com/realm/SwiftLint/pull/3122

**Comment by lordzsolt** [comment]
I'll take a look.

**Comment by lordzsolt** [comment]
I tried fixing it, however I did not find any solution achievable through a simple regex change.

Attempt 1: Require to be preceded by `case`

Regex: `case.*\.\S+\s*(\([,\s_]*\))`

However, this fails when there is more than one case per line: `case .bar↓(_), .bar2↓(_)`

---

Attempt 2: Require at least 1 `_` character between the parenthesis.

Regex: `\.\S+\s*(\([,\s]*[,\s_]+[,\s]*\))`

But this fails on an earlier example `case .bar↓()`

However, the only compiling code which contains `case .bar↓()`,  that I could come up with is:


```swift
enum Foo {
	case bar(Void)
}

let foo: Foo
switch foo {
	case .bar(): break
}
```

Because in case of a Void type, you are allowed write either `.bar()` or `.bar(_)`

Any non-Void type will fail to compile with the error `Tuple pattern cannot match values of the non-tuple type 'T'`

---

Considering I don't think I've ever seen an enum-case with a `Void` type associated value, I would propose removing `case .bar↓()` and `case .bar↓() where method() > 2` from the triggering examples and treat it as a false-negative.

Otherwise, I am not sure how to determine that `.empty()` is a static function, rather than an enum-case. Does the AST contain such information?

**Comment by jpsim** [comment]
> Otherwise, I am not sure how to determine that `.empty()` is a static function, rather than an enum-case. Does the AST contain such information?

Only the post-typechecked AST, which Analyzer Rules operate on, but not standard lint rules.
